🔢 NumPy Tutorial › NumPy ufunc &nbsp;·&nbsp; *Bài 35/44*

[« Simple Arithmetic](034_ufunc_simple_arithmetic.ipynb) &nbsp;•&nbsp; [📚 Mục lục](../../README.md) &nbsp;•&nbsp; [NumPy Logs »](036_ufunc_logs.ipynb)

# Rounding Decimals

> 📖 **Học song song trên W3Schools:** [numpy_ufunc_rounding_decimals.asp](https://www.w3schools.com/python/numpy/numpy_ufunc_rounding_decimals.asp)
>
> 🎯 **Trong bài này:** Rounding Decimals · Truncation · Rounding · Floor · Ceil

## Rounding Decimals — Làm tròn số thập phân

NumPy có 5 cách làm tròn chính:

- truncation (cắt bỏ)
- fix
- rounding (làm tròn)
- floor (làm tròn xuống)
- ceil (làm tròn lên)

## Truncation — Cắt bỏ phần thập phân

Bỏ phần thập phân, trả về số thực gần **0** nhất. Dùng `trunc()` hoặc `fix()`:

In [1]:
import numpy as np

arr = np.trunc([-3.1666, 3.6667])

print(arr)

[-3.  3.]


In [2]:
import numpy as np

arr = np.fix([-3.1666, 3.6667])

print(arr)

[-3.  3.]


## Rounding — Làm tròn

`around()` tăng chữ số trước lên 1 nếu chữ số sau ≥ 5:

In [3]:
import numpy as np

arr = np.around(3.1666, 2)

print(arr)

3.17


## Floor — Làm tròn xuống

`floor()` làm tròn về số nguyên **nhỏ hơn** gần nhất:

In [4]:
import numpy as np

arr = np.floor([-3.1666, 3.6667])

print(arr)

[-4.  3.]


> 📌 `floor()` trả về **số thực**, khác `trunc()` ở chỗ: với số âm, `floor(-3.17)` = `-4`, còn `trunc(-3.17)` = `-3`.

## Ceil — Làm tròn lên

`ceil()` làm tròn về số nguyên **lớn hơn** gần nhất:

In [5]:
import numpy as np

arr = np.ceil([-3.1666, 3.6667])

print(arr)

[-3.  4.]


---

## 🧠 Ghi nhớ nhanh

> - `trunc`/`fix` → bỏ phần thập phân (về phía **0**).
> - `around(x, n)` → làm tròn `n` chữ số.
> - `floor` → làm tròn **xuống** · `ceil` → làm tròn **lên** (với số âm khác `trunc`!).

---

## 📝 Exercise — Bài tập nhanh

**❓ Câu hỏi:** `np.floor(-3.2)` bằng bao nhiêu?

- **A.** `3.0`
- **B.** `-3.2`
- **C.** `-4.0`
- **D.** `-3.0`

<details>
<summary>👉 Xem đáp án</summary>

✅ **Đáp án: C** — `-4.0`

</details>

---

[« Simple Arithmetic](034_ufunc_simple_arithmetic.ipynb) &nbsp;•&nbsp; [📚 Mục lục](../../README.md) &nbsp;•&nbsp; [NumPy Logs »](036_ufunc_logs.ipynb)